crates/in_stream/src/ws.rs

```diff
@@ -0,0 +1,382 @@
+            connect_url,
+        })
+    }
+
+    fn process(&mut self) -> Result<Self::Stream> {
+        match self.state.take() {
+            None => Err(Error::new(ErrorKind::NotFound, "raw stream is None")),
+            Some(state) => match state {
+                PartialWssState::PartialSub(sub) => self.priv_sub(sub),
+                PartialWssState::MidCliHandshake(mid) => {
+                    self.priv_proc_wss_cli_result(mid.handshake())
+                }
+                PartialWssState::MidSrvHandshake(mid) => {
+                    self.priv_proc_wss_srv_result(mid.handshake())
+                }
+                PartialWssState::Ready(stream) => Ok(InStreamWebSocket(stream)),
+            },
+        }
+    }
+}
+
+/// a websocket connection to a remote node
+#[derive(Debug)]
+pub struct InStreamWebSocket<Sub: InStreamPartial>(pub tungstenite::WebSocket<Sub::Stream>);
+
+impl<Sub: InStreamPartial> std::ops::Deref for InStreamWebSocket<Sub> {
+    type Target = tungstenite::WebSocket<Sub::Stream>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<Sub: InStreamPartial> std::ops::DerefMut for InStreamWebSocket<Sub> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<Sub: InStreamPartial> std::convert::AsRef<tungstenite::WebSocket<Sub::Stream>>
+    for InStreamWebSocket<Sub>
+{
+    fn as_ref(&self) -> &tungstenite::WebSocket<Sub::Stream> {
+        &self.0
+    }
+}
+
+impl<Sub: InStreamPartial> std::convert::AsMut<tungstenite::WebSocket<Sub::Stream>>
+    for InStreamWebSocket<Sub>
+{
+    fn as_mut(&mut self) -> &mut tungstenite::WebSocket<Sub::Stream> {
+        &mut self.0
+    }
+}
+
+impl<Sub: InStreamPartial> InStreamFramed for InStreamWebSocket<Sub> {
+    type FrameType = Vec<u8>;
+
+    fn read_frame<T: From<Self::FrameType>>(&mut self) -> Result<T> {
+        match self.0.read_message() {
+            Ok(tungstenite::Message::Binary(data)) => Ok(data.into()),
+            Ok(oth) => Err(Error::new(
+                ErrorKind::InvalidData,
+                format!("unexpected remote message: {:?}", oth),
+            )),
+            Err(tungstenite::error::Error::Io(e)) => Err(e),
+            Err(e) => Err(Error::new(
+                ErrorKind::Other,
+                format!("tungstenite error: {:?}", e),
+            )),
+        }
+    }
+
+    fn write_frame<T: Into<Self::FrameType>>(&mut self, data: T) -> Result<()> {
+        let res = self
+            .0
+            .write_message(tungstenite::Message::Binary(data.into()));
+        match &res {
+            Ok(()) => Ok(()),
+            Err(tungstenite::error::Error::Io(e)) if e.would_block() => {
+                // ignore would-block errors on write
+                // tungstenite queues them in pending, they'll get sent
+                Ok(())
+            }
+            Err(tungstenite::error::Error::Io(_)) => {
+                if let Err(tungstenite::error::Error::Io(e)) = res {
```

Comment:
This looks odd. What are you trying to achieve here?

Reply:
In our version of nightly, you can't do conditional match arms with owned data... you can in stable.

so... I have to use a ref for the top match, but then I need to own the Io error to return it from the function.

Reply:
Second reason I saw today for bumping our rust version to > 1.39.